Topic/2024.0/story doc update (#8)
postsharp/PostSharp.Documentation: conceptual/Introduction/WhatsNew/breaking-changes-20240.md

---
uid: breaking-changes-20240
title: "Breaking Changes in PostSharp 2024.0"
product: "postsharp"
categories: "PostSharp;AOP;Metaprogramming"
---
# Breaking Changes in PostSharp 2024.0

PostSharp 2024.0 contains larger number of breaking changes mainly related to platform upgrades and dependency security.

## PostSharp
* .NET 6.0 SDK and later is required to build .NET Standard, .NET Core and .NET projects. Older SDKs are not supported anymore.
* When building .NET Framework projects, x64 version of PostSharp Compiler will be used by default. 
  Projects that reference x86-specific libraries from projects targeting AnyCPU (default setting) should explicitly set `PostSharpTargetProcessor` property to `x86`.
* When building .NET Framework projects with x86 version of PostSharp Compiler, pipe-server is not used by default anymore. This may lead to longer build times.
  If usage of pipe-server is necessary, `PostSharpUsePipeServer` MSBuild property needs to be set to `true`.

> * PostSharp aspects are now required to use `[PSerializable]` for aspects serialization. 

Reviewer: ```suggestion
* PostSharp aspects are now strongly encouraged to use `[PSerializable]` instead of `[Serializable]` for aspects serialization. 
```

>   Projects that require use of binary serialization or need to continue using `[Serializable]` for aspects, should explicitly set `PostSharpBinaryFormatterAllowed` MSBuild property to `True`.

Reviewer: ```suggestion
  Projects that still need to use `[Serializable]` for aspects should explicitly set `PostSharpBinaryFormatterAllowed` MSBuild property to `True`.
```

> * PostSharp dependency upgrade will lead to package download. When used in environments with restricted access to `nuget.org`, this may require a manual download and deployment of these packages.

Reviewer: I don't understand what this means.

## PostSharp.Patterns.Common
* Requires System.Text.RegularExpressions 4.3.1 (was 4.3.0).

## PostSharp.Patterns.Caching
* The default `BinarySerializer` was changed to `JsonCachingSerializer`. In persisted backends this mean that values are not compatible across with older versions of the library and versions should not be mixed.
* `BinarySerializer` is now considered obsolete (produces a warning).

> * Supports targeting only .NET Framework 4.6.1 and later and .NET Standard 2.0 and later, .NET Core 3.1 and .NET 5.0 and later.

Reviewer: ```suggestion
```

> * Drops targeting support for .NET Core 2.0 to 3.0.

Reviewer: ```suggestion
* Support for targeting .NET Core 3.0 and earlier has been dropped.
```

> * Drops targeting support for .NET Framework 4.5.0 to 4.6.0.

Reviewer: ```suggestion
* Support for targeting .NET Framework 4.6.0 and earlier has been dropped.
```

> * Requires System.Text.Json 6.0.0 (new dependency).

Reviewer: ```suggestion
* There is a new dependency to the `System.Text.Json`  6.0.0  package.
```

## PostSharp.Patterns.Caching.Azure

> * Replaces `AzureCacheInvalidator` and `AzureCacheInvalidator2` with `AzureCacheInvalidator3`, which is based on the latest Azure API.

Reviewer: `AzureCacheInvalidator3` should be renamed to `AzureServiceBusCacheInvalidator`.

> * Supports targeting only .NET Framework 4.7.1 and later and .NET Standard 2.0 and later, .NET Core 3.1 and .NET 5.0 and later.

Reviewer: ```suggestion
```

> * Drops targeting support for .NET Core 2.0 to 3.0.

Reviewer: ```suggestion
* Support for targeting .NET Core 3.0 and earlier has been dropped.
```

> * Drops targeting support for .NET Framework 4.5.0 to 4.7.0.

Reviewer: ```suggestion
* Support for targeting .NET Framework 4.5.0 to 4.7.0 has been dropped.
```

> * Requires Azure.Messaging.ServiceBus, 7.16.1. Depending on target framework this was WindowsAzure.ServiceBus 4.1.2 (.NET Framework) or Microsoft.Azure.ServiceBus 5.2.0 (.NET/.NET Core/.NET Standard). 

Reviewer: ```suggestion
* The package now has a dependency on `Azure.Messaging.ServiceBus` 7.16.1 instead of `WindowsAzure.ServiceBus` or `Microsoft.Azure.ServiceBus` depending on the target framework.
```

* Requires Azure.Identity, 1.10.4.

## PostSharp.Patterns.Diagnostics.AspNetCore

> * Requires `Microsoft.AspNetCore.Mvc.Core` 2.1.0 (was 2.0.0).

Reviewer: ```suggestion
* The dependency to `Microsoft.AspNetCore.Mvc.Core` has been updated from 2.0.0 to 2.1.0.
```

## PostSharp.Patterns.Diagnostics.HttpClient

> * Requires `System.Net.Http` 4.3.4 (was 4.0.0) on .NET Framework.

Reviewer: ```suggestion
* The dependency to `System.Net.Http` has been updated from 4.0.0 to 4.3.4 on .NET Framework.
```

## PostSharp.Patterns.Diagnostics.ApplicationInsights

> * Supports only .NET Framework 4.5.2 and later and .NET Standard 2.0 compatible frameworks.

Reviewer: ```suggestion
```

> * Drops support for .NET Standard 1.3 to 1.6.

Reviewer: ```suggestion
* Support for .NET Standard 1.6 and earlier has been dropped.
```

> * Drops support for .NET Framework 4.5.0 and 4.5.1.

Reviewer: ```suggestion
* Support for .NET Framework 4.5.1 and earlier has been dropped.
```

> * Requires `Microsoft.ApplicationInsights` 2.21.0 (was 2.3.0).

Reviewer: ```suggestion
* The dependency to `Microsoft.ApplicationInsights` has been updated from 2.3.0 to 2.21.0.
```

## PostSharp.Patterns.Diagnostics.Log4Net

> * Requires `log4net` 2.0.12 (was 2.0.0 or 2.0.6 depending on the target framework).

Reviewer: ```suggestion
* The dependency to`log4net` has been updated to 2.0.12 from 2.0.0 or 2.0.6 depending on the target framework.
```

## PostSharp.Patterns.Diagnostics.Microsoft

> * Supports only .NET Standard 2.0 compatible frameworks.

Reviewer: ```suggestion
```

> * Drops support for .NET Standard 1.3 to 1.6.

Reviewer: ```suggestion
* Support for .NET Standard 1.6 and earlier has been dropped.
```

> * Drops support for .NET Framework 4.6.0 to 4.7.0.

Reviewer: ```suggestion
* Support for .NET Framework 4.7.0 and earlier has been dropped.
```

> * Requires `Microsoft.Extensions.Logging` 2.1.0 on all target frameworks.

Reviewer: ```suggestion
* The package has a new dependency to `Microsoft.Extensions.Logging` 2.1.0.
```

## PostSharp.Patterns.Diagnostics.Serilog

> * Requires `Serilog` 2.3.0 on all target frameworks.

Reviewer: ```suggestion
* The dependency to`Serilog` has been updated to 2.3.0.
```
